enhancements/baremetal/enable-baremetal-on-other-platforms.md

```diff
@@ -0,0 +1,186 @@
+---
+title: enable-baremetal-on-other-platforms
+authors:
+  - "@asalkeld"
+  - "@sadasu"
+reviewers:
+  - TBD
+  - "@hardy"
+  - "@romfreiman"
+  - "@dhellman"
+approvers:
+  - TBD
+creation-date: 2021-08-20
+last-updated: 2021-08-20
+status: implementable
+see-also:
+  - "/enhancements/baremetal-provisiong-config.md"
+replaces:
+superseded-by:
+---
+
+# Enable baremetal on other Platforms to support centralized host management
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Baremetal Host API is only available when deploying an OpenShift cluster with the baremetal
+platform (via the IPI or AI (Assisted Installer) workflow). Having the ability to
+manage baremetal hosts from clusters without requiring the cluster to be on baremetal
+would be beneficial to customers.
+
+## Motivation
+
+An initial driver of this feature are the centralized host management use cases
+in edge topologies, which without this feature, is restricted to having the
+central OpenShift cluster deployed on baremetal.
+
+See:
+- https://github.com/openshift/enhancements/blob/master/enhancements/installer/agent-based-installation-in-hive.md
+- https://github.com/openshift/assisted-service/tree/master/docs/hive-integration
+
+### Goals
+
+The specific goals of this proposal are to:
+
+Support the centralized host management use case by partially enabling Baremetal Host API
+on the following platforms:
```

Comment:
AFAIK it would be the same as hub-spoke all on Metal3. CBO will look at the platform purely for validating the Provisioning CR.

Reply:
Normally we run the metal3 pod on the control plane nodes. If the control plane is external, there won't be any control plane nodes. Does that mean metal3 won't schedule, or will it move to worker nodes? If it moves, will the image downloader work in the same way?

Reply:
We have not considered the use case where the control plane is external and metal3 would have to run on worker nodes in this proposal. Currently, metal3 would only run on master nodes. We have limited understanding of this usecase and would like to understand how CBO can distinguish scenarios where metal3 should run on master vs worker nodes. There was also a recent discussion around infrastructure nodes that could be related.

Reply:
If we're worried about avoiding support cases from enabling this feature on the wrong types of clusters, we should make sure that we're covering all of the cases. I don't know the platform type of a hypershift cluster, and I don't see an obvious hypershift type in https://github.com/openshift/api/blob/a6156965faae5ce117e3cd3735981a3fc0e27e27/config/v1/types_infrastructure.go#L193 so I wonder if the type is None?

Reply:
I think hypershift's control-plane is on aws and workers are on "other" - i think normally aws. If they want to be able to deploy metal workers, then we will need to enable on aws? - I suspect that is a while off tho'..
the other general case is the dev-preview of CAPI, but I think the current proposal is a good start at opening up the scope of metal3. WDYT?

Reply:
hypershift development has been AWS-centric, but it does not require AWS.

This proposal is for a use case where a management cluster provisions bare metal hosts and joins them to separate managed clusters. I don't think anyone is (yet?) looking to do that with a management cluster whose control plane is external. Unless/until there are ACM stories being prioritized to run ACM on a hypershift control plane, I think it's out of scope to run metal3 components in a hypershift control plane.

Reply:
> For example, would we expect to run the BMO in a cluster with an external control plane?

I don't see why not. A cluster with an external control plane is just like any other cluster. If the purpose of deploying BMO is to manage baremetal hosts and join them to *other* clusters, I think that is a perfectly valid use case. As a general principle though in hypershift clusters, we don't manage the cluster's nodes within the same cluster. They are managed by the management cluster that is hosting its control plane. There is no reason though for a hypershift cluster not to be a management cluster itself.

> I don't know the platform type of a hypershift cluster, and I don't see an obvious hypershift type in ...

Hypershift is not a platform, and so clusters with platforms of type AWS, None, etc can be hypershift clusters. It's more about the control plane topology, for which we have the `External` value.

> Normally we run the metal3 pod on the control plane nodes. If the control plane is external, there won't be any control plane nodes. Does that mean metal3 won't schedule, or will it move to worker nodes?

Yes, it would have to move to worker nodes if you want to run it on Hypershift clusters. Currently it would "just work" because we are still labeling worker nodes as both masters and workers, but we will get rid of that with 4.9+ clusters.